// @launchdarkly/js-sdk-common/dist/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancelableTimedPromise = exports.timedPromise = exports.sleep = exports.secondsToMillis = exports.shouldRetry = exports.noop = exports.httpErrorMessage = exports.fastDeepEqual = exports.defaultHeaders = exports.deepCompact = exports.debounce = exports.clone = exports.base64UrlEncode = void 0;
const cancelableTimedPromise_1 = require("./cancelableTimedPromise");
Object.defineProperty(exports, "cancelableTimedPromise", { enumerable: true, get: function () { return cancelableTimedPromise_1.cancelableTimedPromise; } });
const clone_1 = require("./clone");
exports.clone = clone_1.default;
const date_1 = require("./date");
Object.defineProperty(exports, "secondsToMillis", { enumerable: true, get: function () { return date_1.secondsToMillis; } });
const debounce_1 = require("./debounce");
exports.debounce = debounce_1.default;
const deepCompact_1 = require("./deepCompact");
exports.deepCompact = deepCompact_1.default;
const fast_deep_equal_1 = require("./fast-deep-equal");
exports.fastDeepEqual = fast_deep_equal_1.default;
const http_1 = require("./http");
Object.defineProperty(exports, "base64UrlEncode", { enumerable: true, get: function () { return http_1.base64UrlEncode; } });
Object.defineProperty(exports, "defaultHeaders", { enumerable: true, get: function () { return http_1.defaultHeaders; } });
Object.defineProperty(exports, "httpErrorMessage", { enumerable: true, get: function () { return http_1.httpErrorMessage; } });
Object.defineProperty(exports, "shouldRetry", { enumerable: true, get: function () { return http_1.shouldRetry; } });
const noop_1 = require("./noop");
exports.noop = noop_1.default;
const sleep_1 = require("./sleep");
exports.sleep = sleep_1.default;
const timedPromise_1 = require("./timedPromise");
exports.timedPromise = timedPromise_1.default;
